crates/bevy_ui/src/layout/mod.rs

```diff
@@ -84,75 +84,108 @@ impl Default for UiSurface {
     }
 }
 
+#[derive(Error, Debug)]
+pub enum SurfaceError {
+    #[error("Couldn't find Tuffy node for entity {0:?}")]
+    NodeNotFound(Entity),
+    #[error("Tuffy Error: {0}")]
+    Tuffy(#[from] taffy::error::TaffyError),
+}
+type SurfaceResult<T> = Result<T, SurfaceError>;
+
 impl UiSurface {
+    fn entity_to_taffy(&self, entity: Entity) -> SurfaceResult<&taffy::node::Node> {
+        self.entity_to_taffy
+            .get(&entity)
+            .ok_or(SurfaceError::NodeNotFound(entity))
+    }
+
     /// Retrieves the Taffy node associated with the given UI node entity and updates its style.
     /// If no associated Taffy node exists a new Taffy node is inserted into the Taffy layout.
-    pub fn upsert_node(&mut self, entity: Entity, style: &Style, context: &LayoutContext) {
+    pub fn upsert_node(
+        &mut self,
+        entity: Entity,
+        style: &Style,
+        context: &LayoutContext,
+    ) -> SurfaceResult<()> {
         let mut added = false;
         let taffy = &mut self.taffy;
         let taffy_node = self.entity_to_taffy.entry(entity).or_insert_with(|| {
             added = true;
-            taffy.new_leaf(convert::from_style(context, style)).unwrap()
+
+            // SAFETY: Despite returning a `Result`, this will never fail.
+            unsafe {
+                taffy
+                    .new_leaf(convert::from_style(context, style))
+                    .unwrap_unchecked()
+            }
         });
 
         if !added {
             self.taffy
-                .set_style(*taffy_node, convert::from_style(context, style))
-                .unwrap();
+                .set_style(*taffy_node, convert::from_style(context, style))?;
         }
-    }
-
-    /// Update the `MeasureFunc` of the taffy node corresponding to the given [`Entity`] if the node exists.
-    pub fn try_update_measure(
-        &mut self,
-        entity: Entity,
-        measure_func: taffy::node::MeasureFunc,
-    ) -> Option<()> {
-        let taffy_node = self.entity_to_taffy.get(&entity)?;
 
-        self.taffy.set_measure(*taffy_node, Some(measure_func)).ok()
+        Ok(())
     }
 
     /// Update the children of the taffy node corresponding to the given [`Entity`].
-    pub fn update_children(&mut self, entity: Entity, children: &Children) {
+    pub fn update_children(&mut self, entity: Entity, children: &Children) -> SurfaceResult<()> {
```

Comment:
I ran into this so many times, thanks for the fix. However, I would like to point out that the panics we get from this is a _symptom_ and not the _cause_ of the issue. Nodes aren't synced well, which leads to the panics and fixing all this *might* hide the problem altogether, leading to inconsistencies in the visible UI. (I had this when the nodes went out of sync, the layout was all whacky, and when a sub-tree was deleted it crashed.)